Crash - aws_cognito_identity_provider when updating idp_identifiers

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version
```
Terraform v1.0.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.45.0
```

### Affected Resource(s)

* aws_cognito_identity_provider

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Debug Output

https://gist.github.com/haavardw/ba266a8b4fcb93b89c08747e99c45832

### Info

- Previous state: One aws_cognito_identity_provider
- New state: Added another aws_cognito_identity_provider and setting `idp_identifiers` for the already existing. 
- After first crash, the new identity provider was created in aws, but did not exist in the state.
- Same crash when doing a `apply -target` on the already existing identity provider.
- Manually resolved the issue by
  - importing the newly created identity provider from aws
  - manually setting the `idp_identifiers` for the already existing identity provider and then `refresh -target` the already existing resource.

```
panic: interface conversion: interface {} is nil, not []interface {}

goroutine 116 [running]:
github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsCognitoIdentityProviderUpdate(0xc001b25b00, 0x63161c0, 0xc001d2e000, 0x0, 0xffffffffffffffff)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-aws/aws/resource_aws_cognito_identity_provider.go:194 +0x63f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc00041d0a0, 0x7c5a368, 0xc001b4afc0, 0xc001b25b00, 0x63161c0, 0xc001d2e000, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/resource.go:352 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00041d0a0, 0x7c5a368, 0xc001b4afc0, 0xc001b40fc0, 0xc001b78900, 0x63161c0, 0xc001d2e000, 0x0, 0x0, 0x0, ...)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00059a438, 0x7c5a368, 0xc001b4afc0, 0xc001b05270, 0xc001b4afc0, 0x6ef0f20, 0xc001b74600)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.6.1/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc00221a720, 0x7c5a410, 0xc001b4afc0, 0xc001b40e70, 0xc00221a720, 0xc001b74690, 0xc00012cba0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x6ef0f20, 0xc00221a720, 0x7c5a410, 0xc001b74690, 0xc001b430e0, 0x0, 0x7c5a410, 0xc001b74690, 0xc001e8a000, 0x7f7)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00147a8c0, 0x7c7bed8, 0xc00142de00, 0xc000152a00, 0xc0007c98c0, 0xb3de840, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc00147a8c0, 0x7c7bed8, 0xc00142de00, 0xc000152a00, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc001404820, 0xc00147a8c0, 0x7c7bed8, 0xc00142de00, 0xc000152a00)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
```

https://github.com/hashicorp/terraform-provider-aws/blob/82ecf28b1f70e0def422c3c1047dc1a4a38ec7f2/aws/resource_aws_cognito_identity_provider.go#L193-L195

--- a/.changelog/19819.txt
+++ b/.changelog/19819.txt
@@ -1,0 +1,3 @@
+```release-note:enhancement
+resource/aws_cognito_identity_provider: Fix updating `idp_identifiers` crash.
+```

--- a/.changelog/19819.txt
+++ b/.changelog/19819.txt
@@ -1,3 +1,3 @@
-```release-note:enhancement
+```release-note:bug
 resource/aws_cognito_identity_provider: Fix updating `idp_identifiers` crash.
 ```

--- a/aws/resource_aws_cognito_identity_provider.go
+++ b/aws/resource_aws_cognito_identity_provider.go
@@ -63,17 +63,10 @@ func resourceAwsCognitoIdentityProvider() *schema.Resource {
 			},
 
 			"provider_type": {
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
-				ValidateFunc: validation.StringInSlice([]string{
-					cognitoidentityprovider.IdentityProviderTypeTypeSaml,
-					cognitoidentityprovider.IdentityProviderTypeTypeFacebook,
-					cognitoidentityprovider.IdentityProviderTypeTypeGoogle,
-					cognitoidentityprovider.IdentityProviderTypeTypeLoginWithAmazon,
-					cognitoidentityprovider.IdentityProviderTypeTypeOidc,
-					cognitoidentityprovider.IdentityProviderTypeTypeSignInWithApple,
-				}, false),
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(cognitoidentityprovider.IdentityProviderTypeType_Values(), false),
 			},
 
 			"user_pool_id": {
@@ -111,7 +104,7 @@ func resourceAwsCognitoIdentityProviderCreate(d *schema.ResourceData, meta inter
 
 	_, err := conn.CreateIdentityProvider(params)
 	if err != nil {
-		return fmt.Errorf("Error creating Cognito Identity Provider: %s", err)
+		return fmt.Errorf("Error creating Cognito Identity Provider: %w", err)
 	}
 
 	d.SetId(fmt.Sprintf("%s:%s", userPoolID, providerName))
@@ -154,15 +147,15 @@ func resourceAwsCognitoIdentityProviderRead(d *schema.ResourceData, meta interfa
 	d.Set("user_pool_id", ip.UserPoolId)
 
 	if err := d.Set("attribute_mapping", aws.StringValueMap(ip.AttributeMapping)); err != nil {
-		return fmt.Errorf("error setting attribute_mapping error: %s", err)
+		return fmt.Errorf("error setting attribute_mapping error: %w", err)
 	}
 
 	if err := d.Set("provider_details", aws.StringValueMap(ip.ProviderDetails)); err != nil {
-		return fmt.Errorf("error setting provider_details error: %s", err)
+		return fmt.Errorf("error setting provider_details error: %w", err)
 	}
 
 	if err := d.Set("idp_identifiers", flattenStringList(ip.IdpIdentifiers)); err != nil {
-		return fmt.Errorf("error setting idp_identifiers error: %s", err)
+		return fmt.Errorf("error setting idp_identifiers error: %w", err)
 	}
 
 	return nil
@@ -191,12 +184,12 @@ func resourceAwsCognitoIdentityProviderUpdate(d *schema.ResourceData, meta inter
 	}
 
 	if d.HasChange("idp_identifiers") {
-		params.IdpIdentifiers = expandStringList(d.Get("supported_login_providers").([]interface{}))
+		params.IdpIdentifiers = expandStringList(d.Get("idp_identifiers").([]interface{}))
 	}
 
 	_, err = conn.UpdateIdentityProvider(params)
 	if err != nil {
-		return fmt.Errorf("Error updating Cognito Identity Provider: %s", err)
+		return fmt.Errorf("Error updating Cognito Identity Provider: %w", err)
 	}
 
 	return resourceAwsCognitoIdentityProviderRead(d, meta)

--- a/aws/resource_aws_cognito_identity_provider_test.go
+++ b/aws/resource_aws_cognito_identity_provider_test.go
@@ -72,6 +72,88 @@ func TestAccAWSCognitoIdentityProvider_basic(t *testing.T) {
 	})
 }
 
+func TestAccAWSCognitoIdentityProvider_idpIdentifiers(t *testing.T) {
+	var identityProvider cognitoidentityprovider.IdentityProviderType
+	resourceName := "aws_cognito_identity_provider.test"
+	userPoolName := fmt.Sprintf("tf-acc-cognito-user-pool-%s", acctest.RandString(7))
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t); testAccPreCheckAWSCognitoIdentityProvider(t) },
+		ErrorCheck:   testAccErrorCheck(t, cognitoidentityprovider.EndpointsID),
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSCognitoIdentityProviderDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSCognitoIdentityProviderIDPIdentifierConfig(userPoolName, "test"),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSCognitoIdentityProviderExists(resourceName, &identityProvider),
+					resource.TestCheckResourceAttr(resourceName, "idp_identifiers.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "idp_identifiers.0", "test"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccAWSCognitoIdentityProviderIDPIdentifierConfig(userPoolName, "test2"),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSCognitoIdentityProviderExists(resourceName, &identityProvider),
+					resource.TestCheckResourceAttr(resourceName, "idp_identifiers.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "idp_identifiers.0", "test2"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSCognitoIdentityProvider_disappears(t *testing.T) {
+	var identityProvider cognitoidentityprovider.IdentityProviderType
+	resourceName := "aws_cognito_identity_provider.test"
+	userPoolName := fmt.Sprintf("tf-acc-cognito-user-pool-%s", acctest.RandString(7))
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t); testAccPreCheckAWSCognitoIdentityProvider(t) },
+		ErrorCheck:   testAccErrorCheck(t, cognitoidentityprovider.EndpointsID),
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSCognitoIdentityProviderDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSCognitoIdentityProviderConfig_basic(userPoolName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSCognitoIdentityProviderExists(resourceName, &identityProvider),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsCognitoIdentityProvider(), resourceName),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
+func TestAccAWSCognitoIdentityProvider_disappears_userPool(t *testing.T) {
+	var identityProvider cognitoidentityprovider.IdentityProviderType
+	resourceName := "aws_cognito_identity_provider.test"
+	userPoolName := fmt.Sprintf("tf-acc-cognito-user-pool-%s", acctest.RandString(7))
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t); testAccPreCheckAWSCognitoIdentityProvider(t) },
+		ErrorCheck:   testAccErrorCheck(t, cognitoidentityprovider.EndpointsID),
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSCognitoIdentityProviderDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSCognitoIdentityProviderConfig_basic(userPoolName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSCognitoIdentityProviderExists(resourceName, &identityProvider),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsCognitoUserPool(), "aws_cognito_user_pool.test"),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
 func testAccCheckAWSCognitoIdentityProviderDestroy(s *terraform.State) error {
 	conn := testAccProvider.Meta().(*AWSClient).cognitoidpconn
 
@@ -193,4 +275,33 @@ resource "aws_cognito_identity_provider" "test" {
   }
 }
 `, userPoolName)
+}
+
+func testAccAWSCognitoIdentityProviderIDPIdentifierConfig(userPoolName, attribute string) string {
+	return fmt.Sprintf(`
+resource "aws_cognito_user_pool" "test" {
+  name                     = %[1]q
+  auto_verified_attributes = ["email"]
+}
+
+resource "aws_cognito_identity_provider" "test" {
+  user_pool_id  = aws_cognito_user_pool.test.id
+  provider_name = "Google"
+  provider_type = "Google"
+
+  idp_identifiers = [%[2]q]
+
+  provider_details = {
+    attributes_url                = "https://people.googleapis.com/v1/people/me?personFields="
+    attributes_url_add_attributes = "true"
+    authorize_scopes              = "email"
+    authorize_url                 = "https://accounts.google.com/o/oauth2/v2/auth"
+    client_id                     = "test-url.apps.googleusercontent.com"
+    client_secret                 = "client_secret"
+    oidc_issuer                   = "https://accounts.google.com"
+    token_request_method          = "POST"
+    token_url                     = "https://www.googleapis.com/oauth2/v4/token"
+  }
+}
+`, userPoolName, attribute)
 }

